### DISCUSSION TITLE
Translatable TagBase subclass

### DISCUSSION THREAD

**Discussion by orencio** [wagtail/wagtail-localize#745] [Q&A]
Hi, I've attempted to make a custom tag model translatable. In concrete I have:

```python
from django.db import models
from taggit.models import TagBase, ItemBase
from wagtail.models import  TranslatableMixin
from wagtail.snippets.models import register_snippet
from wagtail.snippets.views.snippets import SnippetViewSet

class InformationTag(TagBase, TranslatableMixin):
    free_tagging = False

    class Meta:
        verbose_name = 'information tag'
        verbose_name_plural = 'information tags'

        # This is a requisite for 'TranslatableMixin'.
        unique_together = ('translation_key', 'locale')


class InformationPageTag(ItemBase):
    tag = models.ForeignKey(
        InformationTag,
        related_name='information_tag',
        on_delete=models.CASCADE
    )
    content_object = ParentalKey(
        'InformationPage',
        related_name='tagged_items',
        on_delete=models.CASCADE
    )


@register_snippet
class InformationTagManagement(SnippetViewSet):
    panels = [FieldPanel("name")]  # only show the name field
    model = InformationTag
    icon = "tag"  # change as required
    list_display = ["name", "slug"]
    search_fields = ("name",)
```

I've create a new tag from snippets menu and when I try to translate I get this error:

```
DoesNotExist at /admin/localize/submit/snippet/information/informationtag/1/
InformationTag matching query does not exist.
Request Method:	POST
Request URL:	http://localhost:8000/admin/localize/submit/snippet/information/informationtag/1/
Django Version:	4.2.7
Exception Type:	DoesNotExist
Exception Value:	
InformationTag matching query does not exist.
Exception Location:	~/dev/web/wagtail/lib/python3.11/site-packages/django/db/models/query.py, line 637, in get
Raised during:	wagtail_localize.views.submit_translations.SubmitSnippetTranslationView
Python Executable:	~/dev/web/wagtail/bin/python
Python Version:	3.11.6
```

What is happeing?

Are taggit clases translatables?

If they aren't, how is it possible to translate a TagBase subclass?

**Comment by zerolab** [comment]
1. you're using the `@register_snippet` as a decorator in the wrong manner. It needs to go on your model (i.e. `InformationTag`), not the ViewSet - https://docs.wagtail.org/en/stable/topics/snippets/registering.html#using-register-snippet-as-a-decorator
   If you want to use it with your viewset, then use it as a function - `register_snippet(InformationTagManagement)`
2. taggit classes are not translatable since they come from third-party package. 
3. "how is it possible to translate a TagBase subclass" - I am afraid I do not have the capacity to explore that at the moment. In theory, all the required bits are there, but I suspect there's more work to be done to make it happen. If you have the time to dig into the code, a PR will be most welcome, even if it were documentation-only

**Comment by orencio** [comment]
I've seen than when a `TagBase` is going to be translated to other language, in `create_translations` function this error is raised and passed:

```
DEBUG operations.create_translations: ERROR saving translation (e=ValidationError({'name': ['Ya existe Information tag con este Nombre.'], 'slug': ['Ya existe Information tag con este Slug.']})
```

(`name` and `slug` is going to be created and they just exists)

at line  <https://github.com/wagtail/wagtail-localize/blob/a3efa8f65de3445c75ea271b1b0ee513a445deaa/wagtail_localize/operations.py#L86>

This is because both (`name` and `slug`) must be uniques for `InformationTag` due to `TagBase` inherits.

Is it correct?

So I think it is impossible to be a translatable `TagBase` class :-/
